opm/io/eclipse/rst/network.hpp

```diff
@@ -0,0 +1,78 @@
+/*
+  Copyright 2021 Equinor ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify it under the
+  terms of the GNU General Public License as published by the Free Software
+  Foundation, either version 3 of the License, or (at your option) any later
+  version.
+
+  OPM is distributed in the hope that it will be useful, but WITHOUT ANY
+  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+  details.
+
+  You should have received a copy of the GNU General Public License along
+  with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef RST_NETWORK_HPP
+#define RST_NETWORK_HPP
+
+#include <memory>
+#include <optional>
+#include <string>
+#include <vector>
+
+namespace Opm {
+    class UnitSystem;
+} // namespace Opm
+
+namespace Opm { namespace EclIO {
+    class RestartFileView;
+}} // namespace Opm::EclIO
+
+namespace Opm { namespace RestartIO {
+
+    class RstNetwork
+    {
+    public:
+        struct Branch
+        {
+            int down{-1};
+            int up{-1};
+            int vfp{-1};
+        };
+
+        struct Node
+        {
+            std::string name{};
+            std::optional<double> terminal_pressure{};
+            std::optional<std::string> as_choke{};
+            bool add_lift_gas{false};
+        };
+
+        explicit RstNetwork(std::shared_ptr<EclIO::RestartFileView> rstView,
+                            const UnitSystem&                       usys);
+
+        ~RstNetwork();
+        RstNetwork(const RstNetwork& rhs);
+        RstNetwork(RstNetwork&& rhs);
+
+        RstNetwork& operator=(const RstNetwork& rhs);
+        RstNetwork& operator=(RstNetwork&& rhs);
+
+        bool isActive() const;
+
+        const std::vector<Branch>& branches() const;
+        const std::vector<Node>& nodes() const;
+
+    private:
+        class Implementation;
+        std::unique_ptr<Implementation> pImpl_;
```

Comment:
The restart code has been written in a certain pattern - quite different from what you have chosen here. I would appreciate if you would follow the already established pattern

Reply:
> The restart code has been written in a certain pattern - quite different from what you have chosen here. I would appreciate if you would follow the already established pattern

Well, I'm going to argue that approach used here (and for the [aquifers](https://github.com/OPM/opm-common/blob/5e2dc1e30cf18f216f402e87a7eeaf1f863b6e05/opm/io/eclipse/rst/aquifer.hpp#L127-L128) for that matter), is a better way to write it.  This way is also arguably closer to what the restart output code does, not to mention what we already do to initialise the dynamic solution states in `LoadRestart.cpp`.

The only real difference between the two is the way of constructing the representative structures (e.g., [`RstWell`](https://github.com/OPM/opm-common/blob/5e2dc1e30cf18f216f402e87a7eeaf1f863b6e05/opm/io/eclipse/rst/well.hpp#L39)).  Explicit constructors that take restart arrays directly force `RstState::load()` to be aware of every array that could possibly contribute to forming those objects.  Moreover, at least in the current implementation, they are defined in terms of manual and error prone offset calculations that make the expressions more complicated than necessary whilst looping over the subobjects (e.g., wells, segments, groups, or connections).

The approach used in this PR delegates knowledge of how to construct each representative object to that object's component private implementation file and only presents the final results as an iterable sequence.

I will concede that the approach using an explicit constructor is easier to exercise in a unit test context, since the unit test could set up those arrays manually.  On the other hand, we don't use that ability at all.  Every existing unit test that exercises the restart code either loads a file from a known location or calls individual components of the output code in order to form the pertinent arrays from `EclipseState`, `Schedule` and synthetic solution objects as needed.

Reply:
> Well, I'm going to argue that approach used here (and for the aquifers for that matter), is a better way to write it. 

Well - funny thing is that we don't agree here; and then we have to find a solution together.

My main point is about simplicity. The problem at hand is to load and assemble values from various restart vectors, hold on to them for a short time and then pass them on to the schedule objects. During their lifetime this data is immutable, it is not involved in calculation of any sort, the interactions between objects are very limited and the data  is not consumed anywhere outside `Schedule::load_rst()` - i.e. it is simple stuff and the implementation introduces a complexity which only serves to occlude the fact that this is really simple stuff - POD I think they call it.

> at least in the current implementation, they are defined in terms of manual and error prone offset calculations

Personally I like the existing approach better than the windowed array solution - again for clarity, but of course I agree that it is more prone to error. If we are to agree on a compromise of sorts going forward my suggestion would be to use the window approach - for both new and old code - but refrain from pimpl pattern. What do you think?

Reply:
Having looked at the code and discussed the design with @bska and @joakim-hove I agree that using the windowing technique is good for avoiding indexing errors, and should be retained. The pimpl idiom is too much here though, and does not give any significant benefits here in my opinion.

Reply:
> The pimpl idiom is too much here though, and does not give any significant benefits here in my opinion.

I have pushed an update to this effect.